db/document.go

```diff
@@ -38,6 +40,139 @@ const (
 	DocUnmarshalNone                                     // No unmarshalling (skips import/upgrade check)
 )
 
+// Struct populated by an incoming document
+type IncomingDocument struct {
+	BodyBytes         []byte // BodyBytes contains the raw document only. Doesn't include any special properties
+	Body              Body   // Body contains the marshalled document only. Doesn't include any special properties until syncFn is generated.
+	SpecialProperties        // Embedded SpecialProperties struct. Extracted either from body or can be manually entered
+}
+
+// Stores special properties which are a part of an incoming document
+type SpecialProperties struct {
+	DocID       string
+	RevID       string
+	Deleted     bool
+	Expiry      uint32
+	Attachments AttachmentsMeta
+	Revisions   Revisions
+}
+
+// Simply constructs IncomingDocument struct with provided body bytes. This is only used to enforce that body is not
+// directly modified.
+func NewIncomingDocument(body []byte) *IncomingDocument {
+	return &IncomingDocument{
+		BodyBytes:         body,
+		SpecialProperties: SpecialProperties{},
+	}
+}
+
+// Must ensure that the returned body here is never mutated. In the case where we just return body with stamped in
+// special properties we need to ensure nothing else mutates body after this point, otherwise bytes and body will
+// become disconnected and reflect different values.
+func (doc *IncomingDocument) GetSyncFnBody() (map[string]interface{}, error) {
+	var syncFnBody map[string]interface{}
+
+	if doc.Body == nil {
+		buf := bytes.NewBuffer(doc.BodyBytes)
+		d := json.NewDecoder(buf)
+		d.UseNumber()
+		err := d.Decode(&doc.Body)
+		if err != nil {
+			return nil, err
+		}
+	}
+	syncFnBody = doc.Body
+	stampSyncFnProperties(syncFnBody, doc.SpecialProperties)
```

Comment:
I'm not sure this is a situation we could be in (maybe for the CAS retry?) - but what happens when `GetSyncFnBody()` is called twice with different properties (rev ID?)?

We'll be trying to stamp properties twice, but the logic inside `stampSyncFnProperties` skips updating properties if they're already present. This might get you into a situation where it's not possible to update a sync function body with the current set of properties.

Reply:
Does it? Pretty sure it would end up with the new correct body in the case where special properties have been updated